File: leetcode/linkedlist/203.Remove-Linked-List-Elements/203.Remove-Linked-List-Elements.go
package leetcode

import (
	"github.com/yuancf1024/algorithm-go/structures"
)

type ListNode = structures.ListNode

/*
给你一个链表的头节点 head 和一个整数 val ，请你删除链表中所有满足 
Node.val == val 的节点，并返回 新的头节点 。
*/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// // 2022-04-13 自己写的
// func removeElements(head *ListNode, val int) *ListNode {
//     if head == nil {
//         return nil
//     }

//     preHead := &ListNode{Val: 0}
//     cur := preHead // 这里没有把cur 和 head联系起来
//     cur := head

//     for cur != nil {
//         if cur.Val == val {
//             cur.Next = cur.Next.Next // 犯错点：没有把preHead 和 cur 连接起来，而PreHead又不能直接使用
//             // 因此需要使用一个中间指针过渡一下
//         }
//         cur = cur.Next
//     }

//     return preHead.Next
// }

func removeElements(head *ListNode, val int) *ListNode {
    if head == nil {
        return head
    }

    newHead := &ListNode{Val: 0, Next: head}
    pre := newHead
    cur := head
    for cur != nil {
        if cur.Val == val {
            pre.Next = cur.Next // 跳过cur节点
        } else {
            pre = cur
        }
        cur = cur.Next
    }
    return newHead.Next
}

File: leetcode-cpp/203.Remove-Linked-List-Elements/203.Remove-Linked-List-Elements.cpp
/* 203. Remove Linked List Elements
给你一个链表的头节点 head 和一个整数 val ，
请你删除链表中所有满足 Node.val == val 的节点，
并返回 新的头节点 。
 
示例 1：
输入：head = [1,2,6,3,4,5,6], val = 6
输出：[1,2,3,4,5]

示例 2：
输入：head = [], val = 1
输出：[]

示例 3：
输入：head = [7,7,7,7], val = 7
输出：[]
 
提示：

列表中的节点数目在范围 [0, 10^4] 内
1 <= Node.val <= 50
0 <= val <= 50
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// 方法1：直接使用原来的链表进行移除节点操作
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        // 删除头节点
        while (head != NULL && head->val == val) { // 注意这里不是if
            ListNode* tmp = head;
            head = head->next;
            delete tmp;
        }

        // 删除非头节点
        ListNode* cur = head;
        while (cur != NULL && cur->next != NULL) {
            if (cur->next->val == val) {
                ListNode* tmp = cur->next;
                cur->next = cur->next->next;
                delete tmp;
            } else {
                cur = cur->next;
            }
        }
        return head;
    }
};

// 方法2：设置一个虚拟头节点再进行移除节点操作
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* dummyHead = new ListNode(0); // 设置一个虚拟头节点
        dummyHead->next = head; // 将虚拟头节点指向head，这样方便后面做删除操作
        ListNode* cur = dummyHead;
        while (cur->next != NULL) {
            if (cur->next->val == val) {
                ListNode* tmp = cur->next;
                cur->next = cur->next->next;
                delete tmp;
            } else {
                cur = cur->next;
            }
        }
        head = dummyHead->next;
        delete dummyHead;
        return head;
    }
};

/*
解题思路：
链表的移除操作：移除操作，就是让节点next指针直接指向下下一个节点就可以了。

- 方式1：直接使用原来的链表来进行删除操作。
- 方式2：设置一个虚拟头结点在进行删除操作
*/